src/query/api/v1/handler/database/create.go

```diff
@@ -124,52 +142,77 @@ func NewCreateHandler(
 	cfg config.Configuration,
 	embeddedDbCfg *dbconfig.DBConfiguration,
 ) http.Handler {
+	placementHandlerOptions := placement.HandlerOptions{
+		ClusterClient: client,
+		Config:        cfg,
+	}
 	return &createHandler{
-		placementInitHandler: placement.NewInitHandler(
-			placement.HandlerOptions{ClusterClient: client, Config: cfg}),
+		placementInitHandler:   placement.NewInitHandler(placementHandlerOptions),
+		placementGetHandler:    placement.NewGetHandler(placementHandlerOptions),
 		namespaceAddHandler:    namespace.NewAddHandler(client),
+		namespaceGetHandler:    namespace.NewGetHandler(client),
 		namespaceDeleteHandler: namespace.NewDeleteHandler(client),
 		embeddedDbCfg:          embeddedDbCfg,
 	}
 }
 
 func (h *createHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	ctx := r.Context()
-	logger := logging.WithContext(ctx)
+	var (
+		ctx    = r.Context()
+		logger = logging.WithContext(ctx)
+	)
 
-	namespaceRequest, placementRequest, rErr := h.parseRequest(r)
+	currPlacement, _, err := h.placementGetHandler.Get(placement.M3DBServiceName, nil)
+	if err != nil {
+		logger.Error("unable to get  placement", zap.Error(err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	parsedReq, namespaceRequest, placementRequest, rErr := h.parseAndValidateRequest(r, currPlacement)
 	if rErr != nil {
 		logger.Error("unable to parse request", zap.Any("error", rErr))
 		xhttp.Error(w, rErr.Inner(), rErr.Code())
 		return
 	}
 
-	nsRegistry, err := h.namespaceAddHandler.Add(namespaceRequest)
+	currPlacement, badRequest, err := h.maybeInitPlacement(currPlacement, parsedReq, placementRequest, r)
 	if err != nil {
-		logger.Error("unable to add namespace", zap.Any("error", err))
-		xhttp.Error(w, err, http.StatusInternalServerError)
+		logger.Error("unable to initialize placement", zap.Error(err))
+		status := http.StatusBadRequest
+		if !badRequest {
+			status = http.StatusInternalServerError
+		}
+		xhttp.Error(w, err, status)
 		return
 	}
 
-	initPlacement, err := h.placementInitHandler.Init(placement.M3DBServiceName, r, placementRequest)
+	nsRegistry, err := h.namespaceGetHandler.Get()
 	if err != nil {
-		// Attempt to delete the namespace that was just created to maintain idempotency.
-		nsDeleteErr := h.namespaceDeleteHandler.Delete(namespaceRequest.Name)
-		if nsDeleteErr != nil {
-			logger.Error(
-				"unable to delete namespace we just added",
-				zap.Any("originalError", err),
-				zap.Any("namespaceDeleteError", nsDeleteErr))
-			xhttp.Error(w, err, http.StatusInternalServerError)
-			return
-		}
+		logger.Error("unable to retrieve existing namespaces", zap.Error(err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
 
-		logger.Error("unable to initialize placement", zap.Any("error", err))
+	// TODO(rartoul): Add test for NS exists.
+	_, nsExists := nsRegistry.Namespaces[namespaceRequest.Name]
```

Comment:
Is this kind of racy?

Reply:
discussed offline, its racy but we're just trying to provide a good error message here, the namespace API itself won't allow you to overwrite an existing namespace